Allow logging configuration

grahambrereton-form3 opened the pull request:
<!--
Thank you for contributing to Chaos Mesh!

If you're unsure where to start, please refer to the contributing doc:

https://github.com/chaos-mesh/chaos-mesh/blob/master/CONTRIBUTING.md

If you still have questions, please let us know via issues.

Please follow https://www.conventionalcommits.org/en/v1.0.0/ when you open a new PR:
-->

## What problem does this PR solve?

Allow configuration of key logging parameters through environment, and expose these environment variables through the helm chart values.

Allow configuration for truncating log fields over a certain limit. This prevents log aggregation failure in some cases where very large objects are logged.

Preserves similar behaviour to present if no configuration is provided.

<!-- Uncomment this line if some issues to close -->
<!-- Close #<issue number> -->

## What's changed and how it works?

New configuration environment variables added to setup zap logger in a configurable way.

These environment variables are then exposed in the helm chart through a new top-level key `logging` so that the logging config can be provided in one place and applied to all deployed applications.

<!-- Uncomment this line if this PR is associated with a proposal -->
<!-- Proposal: [name](url) -->

## Related changes

- [ ] This change also requires further updates to the [website](https://github.com/chaos-mesh/website) (e.g. docs)
- [ ] This change also requires further updates to the `UI interface`

## Cherry-pick to release branches (optional)

> This PR should be cherry-picked to the following release branches:

- [ ] release-2.6
- [ ] release-2.5

## Checklist

### CHANGELOG

> Must include at least one of them.

- [x] I have updated the `CHANGELOG.md`
- [ ] I have labeled this PR with "no-need-update-changelog"

### Tests

> Must include at least one of them.

- [x] Unit test
- [ ] E2E test
- [ ] Manual test

### Side effects

- [ ] **Breaking backward compatibility**

## DCO

If you find the DCO check fails, please run commands like below (Depends on the actual situations. For example, if the failed commit isn't the most recent) to fix it:

```shell
git commit --amend --signoff
git push --force
```

grahambrereton-form3 commented:
I tried running the failing unit tests locally, and they passed. Not sure what caused the failure in CI.

STRRL commented on pkg/log/zap.go:
```diff
@@ -16,26 +16,140 @@
+		var err error
+		logLevel, err = zapcore.ParseLevel(envLevel)
+		if err != nil {
+			return logr.Discard(), err
```
when we fail to parse the loglevel, please use the default log level, and just use fmt.Println to print a message about this error and notice user we fallback to the default log level.

STRRL commented on pkg/log/zap.go:
```diff
@@ -16,26 +16,140 @@
+			var timeEncoder zapcore.TimeEncoder
+
+			if err := timeEncoder.UnmarshalText([]byte(v)); err != nil {
+				return logr.Discard(), err
```
ditto, please fallback to a default timestamp format if we failed to parse that from ENV

grahambrereton-form3 replied:
Interesting detail here is that `timeEncoder.UnmarshalText` in its current implementation never returns an error and silently defaults to `epoch`. I've added the error handler that just prints a warning, because it seems a bad idea to leave the error completely unhandled, but it won't be called with the current implementation in zap.

Could potentially drop the `else` clause that prints the error, given it's currently unreachable.

STRRL commented:
Hi @grahambrereton-form3, thanks for your contribution!

because the logger is a kind of application basic infrastructure so we'd better make it more robust.

grahambrereton-form3 commented:
@STRRL I pushed a new commit, which should address your concerns.

miketonks-form3 commented:
Adding my review: LGTM.  

We've been using this in our local build for some time, and works well.  We export logs via fluentd to centralised log store, and the structured logging is a nice improvement.

grahambrereton-form3 committed:
feat: allow logger configuration

Configure structured logging and conventional field names for timestamp,
level, etc.

Allow configuration of key logging parameters through environment.

Truncate log fields over a certain limit.

Co-authored-by: Kenneth Buck <kenneth.buck@form3.tech>
Signed-off-by: Graham Brereton <graham.brereton@form3.tech>

grahambrereton-form3 commented:
Rebased on master and reset author to me on the commit to make DCO happy :crossed_fingers:

patsevanton commented:
I would really like to manage the logging level. Can I marge this pull request?

STRRL commented:
> I would really like to manage the logging level. Can I marge this pull request?

Yes I think we could have that. let me rebase it and do some updates;

patsevanton commented:
Friendly reminder about this PR.

STRRL commented:
this branch does not allow me to push it, I am going to work in another branch
